chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,338 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{error::ConnectTransactionError, TransactionSource};
+
+use common::{
+    chain::{Block, PoolId},
+    primitives::{Amount, Id},
+    Uint256,
+};
+use pos_accounting::{AccountingBlockRewardUndo, PoSAccountingOperations, PoSAccountingView};
+
+use super::accounting_delta_adapter::PoSAccountingDeltaAdapter;
+
+/// Distribute reward among the staker and delegators
+pub fn distribute_reward<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_reward: Amount,
+) -> Result<AccountingBlockRewardUndo, ConnectTransactionError> {
+    let tx_source = TransactionSource::Chain(block_id);
+
+    let pool_data = accounting_adapter
+        .accounting_delta()
+        .get_pool_data(pool_id)?
+        .ok_or(ConnectTransactionError::PoolDataNotFound(pool_id))?;
+
+    let total_staker_reward = (total_reward - pool_data.cost_per_block())
+        .and_then(|v| (v * pool_data.margin_ratio_per_thousand().into()).and_then(|v| v / 1000))
+        .and_then(|v| v + pool_data.cost_per_block())
+        .ok_or(ConnectTransactionError::StakerRewardCalculationFailed(
+            block_id,
+        ))?;
+    let total_delegators_reward = (total_reward - total_staker_reward).ok_or(
+        ConnectTransactionError::DelegatorRewardCalculationFailed(block_id),
+    )?;
+    let total_delegators_reward = Uint256::from_amount(total_delegators_reward);
+
+    let increase_balance_undo = accounting_adapter
+        .operations(tx_source)
+        .increase_pool_balance(pool_id, total_staker_reward)?;
+
+    let delegation_undos = accounting_adapter
+        .accounting_delta()
+        .get_pool_delegations_shares(pool_id)?
+        .map(|delegation_shares| {
+            let total_delegators_balance =
+                delegation_shares.values().try_fold(Amount::ZERO, |acc, v| {
+                    (acc + *v).ok_or(ConnectTransactionError::DelegatorsRewardSumFailed(block_id))
+                })?;
+            let total_delegators_balance = Uint256::from_amount(total_delegators_balance);
+
+            delegation_shares
+                .iter()
+                .map(|(delegation_id, balance)| {
+                    let balance = Uint256::from_amount(*balance);
+                    let reward = total_delegators_reward * balance / total_delegators_balance;
+                    let reward: u128 = reward.try_into().map_err(|_| {
+                        ConnectTransactionError::DelegatorRewardCalculationFailed(block_id)
+                    })?;
+
+                    accounting_adapter
+                        .operations(tx_source)
+                        .delegate_staking(*delegation_id, Amount::from_atoms(reward))
```

Comment:
Keep in mind that since you're using division and fractional arithmetic with integers, it's not always correct that the sum of all amounts is equal to the total.

An easy example is dividing 4 among shares of 1:1:1. That's basically 4/3+4/3+4/3 = 3 which is not 4, which means you'll have 1 unit undistributed. Did you account for that?

I'm not saying you have to hand it to someone, but you have to define the protocol parameters, by which such "burning" is acceptable. You're basically burning MLT for this. 

What happens if the amounts are so small that you got some zeros?

Usually the better way to do this (from an integer arithmetic mathematical point of view) is to do the distribution in memory in some data structure, confirm the constraints, and then write it to your cache/deltas. This conforms to the read-modify-write paradigm. Though the cost here is a temporary allocation of a relatively large data structure in memory.

You can't have it both ways. Either pre-calculate, do your constraints verification and ensure that the amounts are sane, OR, have this kind of vulnerabilities where integer arithmetic policy is unclear.

The alternative is some ugly, mutable data structure that changes while you're writing on disk. I don't prefer that.

Reply:
Do you suggest another approach instead of burning or is it the implementation that bothers you?

For me, it's either we figure out what to do with the remainder (burn, tip owner, give to a delegator) or play with fixed-point balances.

Reply:
Basically what I'm saying is that with such algorithm we should be careful because of corner cases. I suggest moving the numeric distribution function to a separate function that can be tested and where corner cases are verified.

Maybe giving the remainder to the staker is OK. I don't see another way.

Reply:
Now all the remainder goes to the owner. Also if there are delegations but their balance is 0 the reward also goes to the owner. Tests included.